src/lib.rs

```diff
@@ -256,35 +256,30 @@ pub mod contract {
                     return Err(InvariantError::PriceLimitReached);
                 }
 
-                // TODO: refactor
-                let mut tick = Tick::default();
-
-                let update_limiting_tick = limiting_tick.map(|(index, bool)| {
-                    if bool {
-                        tick = self.ticks.get(pool_key, index).unwrap();
-                        (index, Some(&mut tick))
-                    } else {
-                        (index, None)
+                if let Some((tick_index, is_initialized)) = limiting_tick {
+                    if is_initialized {
+                        let mut tick = self.ticks.get(pool_key, tick_index).unwrap();
+                        let has_crossed = pool.cross_tick(
+                            result,
+                            swap_limit,
+                            tick_index,
+                            &mut tick,
+                            &mut remaining_amount,
+                            by_amount_in,
+                            x_to_y,
+                            current_timestamp,
+                            &mut total_amount_in,
+                            self.state.protocol_fee,
+                            pool_key.fee_tier,
+                        );
+                        if has_crossed {
+                            self.emit_cross_tick_event(caller, pool_key, tick_index);
```

Comment:
emitting events should be outside `calculate_swap` function. Now even quote message emitting cross events.